### ISSUE TITLE
generic hook not available for expl3 commands if the replacement text contains :~ 

### ISSUE THREAD

**Issue by u-fischer** [latex3/latex2e#1099]
## Brief outline of the bug

In the following example one gets and error if one try to use the generic hook of `\testB`:

## Minimal example showing the bug

```latex
\RequirePackage{latexbug}    
\documentclass{article}
\ExplSyntaxOn %required
\def\testA#1{hallo:}
\def\testB#1{hallo:~}

\AddToHook{cmd/testA/before}{\tiny}
\AddToHook{cmd/testB/before}{\tiny}

\ExplSyntaxOff

\begin{document}
blub
\end{document}
```

gives 

~~~~
! LaTeX hooks Error: Generic hooks cannot be added to '\testB'.
~~~~

## Log file (required) and possibly PDF file  

[test-utf8.log](https://github.com/latex3/latex2e/files/11860052/test-utf8.log)

**Comment by FrankMittelbach** [comment]
Well, the original example fails because of the `:` having catcode 11 and later on catcode 12 when the patching happens.

The problem here seems to me that something like `hello:~` is actually wrong as it produces the word `hello:` not `hello` followed by the punctuation `:`  (which is what is intended I would assume if a command contains text to be used later on). And expl3 doesn't offer a solution for this and I'm not sure how it could easily. 

We can't use `\:` because that has a meaning in text. We could perhaps use `\?`  or `\<` but that wouldn't be pretty and furthermore it would likely break some documents that also discovered that these control symbols aren't in use.
We could offer `\textcolon` but that has the disadvantage that you need to write `\textcolon\space` most of the time.

@PhelypeOleinik any thoughts?

If we aren't going in that direction but accept that we end up with catcode 11 colons in text inside `\ExplSyntaxOn` then perhaps the patching could make a second pass before giving up. If it can't patch it could try to patch once more with  `:` having catcode 11.

**Comment by u-fischer** [comment]
> Well, the original example fails because of the `:` having catcode 11 and later on catcode 12 when the patching happens

Yes something like this. What still eludes me, is why patching a definition with replacement text `hallo:world` and `hallo~world`  works, but patching fails, if both are present `hallo~world:earth`.

**Comment by FrankMittelbach** [comment]
as a guess: because the moment the definition contains spaces a different patching code is used?

**Comment by PhelypeOleinik** [comment]
In Ulrike's original example, patching `\testA` works because it contains a `:`<sub>11</sub>, so the patching code detects (more or less by trial-and-error) that pathching with `\ExplSyntaxOn` works to patch without changing the token list.

`\testB`, however, is written as `hallo:~`, but the `:~` is tokenised as `:`<sub>11</sub>` `<sub>10</sub>.  The problem is that when trying to retokenise, the code fails to retokenise with `\ExplSyntaxOff` because the `:`<sub>11</sub> is re-read as `:`<sub>12</sub>, and it fails to retokenise with `\ExplSyntaxOn` because the ` `<sub>10</sub> is ignored, so the code can't find the correct catcode regime to patch.

It just occurred to me that the problem is that `~`<sub>10</sub> is tokenised as ` `<sub>10</sub>, then ` `<sub>10</sub> is retokenised as ` `<sub>9</sub>, so instead of using "plain `\ExplSyntaxOn`", the right approach would be "`\ExplSyntaxOn` but spaces aren't ignored".  So this might work:
```
\cs_gset_protected:Npn \__hook_retokenize_patch:Nnn #1 #2 #3
  {
    \str_if_eq:eeTF { \cs_argument_spec:N #1 } { }
      { \__hook_patch_expand_redefine:NNnn \c_false_bool #1 {#2} {#3} }
      {
        \__hook_patch_debug:x { ..~command~can~only~be~patched~by~rescanning }
        \tl_set:Nx \l__hook_tmpa_tl
          {
            \int_compare:nNnTF { \char_value_catcode:n {`\@ } } = { 12 }
              { \exp_not:N \makeatletter } { \exp_not:N \makeatother }
          }
        \tl_set:Nx \l__hook_tmpb_tl
          {
            \bool_if:NTF \l__kernel_expl_bool
              { \ExplSyntaxOff }
              {
                \ExplSyntaxOn
                \char_set_catcode_space:n { 32 } % <<<<<<<<<<<<<<<<<<
              }
          }
        \use:x
          {
            \exp_not:N \__hook_try_patch_with_catcodes:Nnnnw
                \exp_not:n { #1 {#2} {#3} }
              { \prg_do_nothing: }
              { \exp_not:V \l__hook_tmpa_tl } % @
              { \exp_not:V \l__hook_tmpb_tl } % _:
              {
                \exp_not:V \l__hook_tmpa_tl   % @
                \exp_not:V \l__hook_tmpb_tl   % _:
              }
          }
              \q_recursion_tail \q_recursion_stop
          {
            \msg_error:nnxx { hooks } { cant-patch }
              { \c_backslash_str #2 } { retok }
          }
      }
  }
```

**Comment by FrankMittelbach** [comment]
So there are in fact 2 code branches and the second attempt wasn't quite right. @PhelypeOleinik where is that patching happening? I thought it only happens in one place during \document and at that point \ExplSyntaxOn would be the wrong attempt anyway, wouldn't it? I mean ~ shouldn't be a space during that rescan nor should space be ignored and perhaps \ExplSyntaxOn does something else that should happen then. Basically one would need to try with the catcodes that can actually end up in the macro code defined and not exactly the ones that \ExplSyntaxOn sets.  And what about the branch that sets \ExplSyntaxOff? Can that ever happen?

(I must confess I haven't looked at the code except for the above, so I may ask nonsense questions).
